library/alloc/src/vec/drain.rs

```diff
@@ -108,9 +121,11 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
 
         impl<'r, 'a, T, A: Allocator> Drop for DropGuard<'r, 'a, T, A> {
             fn drop(&mut self) {
-                // Continue the same loop we have below. If the loop already finished, this does
-                // nothing.
-                self.0.for_each(drop);
+                if mem::needs_drop::<T>() {
```

Comment:
Last time I did something similar this it triggered a compiler bug in a single 3rd-party crate (according to the crater run at that time). #78139
The solution was to just remove the uses of `needs_drop` since they only were an optimization my case. No attempt was made to fix the underlying bug, so it may still be there. But on the other hand your change would affect different iterators and so might not trigger it.

That said `needs_drop` is stable and safe, so imo this PR shouldn't be blocked on a compiler bug of uncertain existence or relevance.

Perhaps a try run + `rustup-toolchain-install-master` + using that to compile the crate in question could shed some light on it.

Reply:
Actually, if `needs_drop() == false` is needed for correctness then this is invalid anyway, [its documentation](https://doc.rust-lang.org/std/mem/fn.needs_drop.html) clearly says so

> This is purely an optimization hint, and may be implemented conservatively: it may return true for types that don't actually need to be dropped. As such **always returning true would be a valid implementation of this function**.

Reply:
That could be solved by a helper Trait that specializes on `T: Copy`

Reply:
> Actually, if needs_drop() == false is needed for correctness then this is invalid anyway

It's not needed for correctness here, but without going via `TrustedRandomAccess` is rather inefficient because that wouldn't advance the inner slice iterator here so on `drop()` first the slice iterator would be advanced over the whole range for no good reason, while without `TrustedRandomAccess` it would already be at the end anyway (unless `Drain` was dropped before it finished).

Or did I misunderstand what you mean?

Reply:
Ok, then only the compiler bug concern applies.

Reply:
Yeah, will give that a try later but maybe this means that this should get a crater run (if that crate builds fine now).

Reply:
Builds without problems